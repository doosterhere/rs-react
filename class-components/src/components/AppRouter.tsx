import { createBrowserRouter, RouterProvider } from 'react-router-dom';

import { ROUTES } from '../constants';

import { HomePage, Page404 } from '../pages';
import { ListItemDetailed } from './ListItemDetailed/ListItemDetailed';

const router = createBrowserRouter([
  {
    path: ROUTES.HOME,
    element: <HomePage />,
    errorElement: <Page404 />,
    children: [
      {
        path: ROUTES.DETAIL,
        element: <ListItemDetailed />,
      },
    ],
  },
]);

const AppRouter = () => <RouterProvider router={router} />;

export { AppRouter };
