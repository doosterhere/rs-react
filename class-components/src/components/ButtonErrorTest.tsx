import { useState } from 'react';

const ButtonErrorTest = () => {
  const [hasError, setHasError] = useState(false);

  const handleThrowError = () => {
    setHasError(true);
  };

  if (hasError) {
    throw new Error('Throwing the error while rendering');
  }

  return (
    <>
      <button type="button" className="button-error" onClick={handleThrowError}>
        Throw an error
      </button>
    </>
  );
};

export default ButtonErrorTest;
