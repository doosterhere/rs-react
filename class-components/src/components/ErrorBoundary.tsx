import { Component, ReactNode } from 'react';

interface IErrorBoundaryProps {
  children: ReactNode;
}

class ErrorBoundary extends Component<IErrorBoundaryProps> {
  state = {
    hasError: false,
    errorMessage: '',
  };

  static getDerivedStateFromError(error: Error) {
    return { hasError: true, errorMessage: error.message.toString() };
  }

  componentDidCatch(error: Error, errorInfo: React.ErrorInfo) {
    console.error(`Error in error boundary: ${error}`, `Component stack: ${errorInfo.componentStack}`);
  }

  reloadPage() {
    window.location.assign('/');
  }

  render() {
    return this.state.hasError ? (
      <div className="container" style={{ display: 'flex', alignItems: 'center' }}>
        <h1 data-testid="error-boundary">Something went wrong!</h1>
        <button onClick={this.reloadPage} style={{ width: '300px' }}>
          Return to main page
        </button>
      </div>
    ) : (
      this.props.children
    );
  }
}

export { ErrorBoundary };
