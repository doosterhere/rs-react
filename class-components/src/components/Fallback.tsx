const Fallback = () => {
  const reloadPage = () => {
    history.go(0);
  };

  return (
    <div>
      <h1>Something went wrong!</h1>
      <button onClick={reloadPage}>Return to main page</button>
    </div>
  );
};

export default Fallback;
