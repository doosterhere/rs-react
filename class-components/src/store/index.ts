export * from './store';
export * from './reducers/itemsReducer';
