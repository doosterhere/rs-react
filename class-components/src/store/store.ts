import { combineReducers, configureStore } from '@reduxjs/toolkit';

import itemsReducer from './reducers/itemsReducer';

const rootReducer = combineReducers({
  items: itemsReducer,
});

const configStore = () =>
  configureStore({
    reducer: rootReducer,
    middleware: getDefaultMiddleware => getDefaultMiddleware({}),
  });

export { configStore };
export type AppStore = ReturnType<typeof configStore>;
export type RootState = ReturnType<typeof rootReducer>;
export type AppDispatch = AppStore['dispatch'];
