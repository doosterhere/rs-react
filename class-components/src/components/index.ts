import SearchBar from './SearchBar.tsx';
import List from './List.tsx';
import ListItem from './ListItem.tsx';
import Loader from './Loader.tsx';
import ErrorBoundary from './ErrorBoundary.tsx';
import ButtonErrorTest from './ButtonErrorTest.tsx';
import Fallback from './Fallback.tsx';
import Modal from './Modal.tsx';
import AppRouter from './AppRouter.tsx';
import Layout from './Layout.tsx';

export {
  SearchBar,
  List,
  ListItem,
  Loader,
  ErrorBoundary,
  ButtonErrorTest,
  Fallback,
  Modal,
  AppRouter,
  Layout,
};
