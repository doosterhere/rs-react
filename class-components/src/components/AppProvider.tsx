'use client';

import { ReactNode, useRef } from 'react';
import { Provider } from 'react-redux';

import '../styles/global.css';

import { AppStore, configStore } from '../store/store';
import { ThemeProvider } from '../components/ThemeContext';
import { ErrorBoundary } from './ErrorBoundary';

function AppProvider({ children }: { children: ReactNode }) {
  const storeRef = useRef<AppStore>();

  if (!storeRef.current) {
    storeRef.current = configStore();
  }

  return (
    <Provider store={storeRef.current}>
      <ErrorBoundary>
        <ThemeProvider>{children}</ThemeProvider>
      </ErrorBoundary>
    </Provider>
  );
}

export { AppProvider };
