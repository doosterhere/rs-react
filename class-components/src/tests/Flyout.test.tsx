import 'whatwg-fetch';
import { screen, render } from '@testing-library/react';
import * as reduxHooks from 'react-redux';
import userEvent from '@testing-library/user-event';

import { Flyout } from '../components';
import { clearSelectedItems } from '../store';

jest.mock('react-redux');

describe('Flyout', () => {
  it('should render correctly', () => {
    jest.spyOn(reduxHooks, 'useSelector').mockReturnValue('1');

    const { container } = render(<Flyout />);

    expect(container).toMatchSnapshot();

    expect(screen.getByText('Selected items: 1')).toBeInTheDocument();
    expect(screen.getAllByRole('button')).toHaveLength(2);
  });

  it('should show and hide confirmation dialog', async () => {
    jest.spyOn(reduxHooks, 'useSelector').mockReturnValue('1');

    render(<Flyout />);

    const unselect = screen.getByRole('button', { name: 'Unselect all' });

    await userEvent.click(unselect);

    expect(screen.getByText('Are you sure?')).toBeInTheDocument();

    const cancel = screen.getByRole('button', { name: 'No way!' });

    await userEvent.click(cancel);

    expect(screen.queryByText('Selected itesms: 1')).not.toBeInTheDocument();
  });

  it('should call dispatcher with proper action to unselect all items', async () => {
    const dispatch = jest.fn();
    jest.spyOn(reduxHooks, 'useSelector').mockReturnValue('1');
    const mockedDispatch = jest.spyOn(reduxHooks, 'useDispatch');
    mockedDispatch.mockReturnValue(dispatch);

    render(<Flyout />);

    const unselect = screen.getByRole('button', { name: 'Unselect all' });

    await userEvent.click(unselect);

    expect(screen.getByText('Are you sure?')).toBeInTheDocument();

    const confirm = screen.getByRole('button', { name: 'Absolutely!' });

    await userEvent.click(confirm);

    expect(dispatch).toHaveBeenCalledWith(clearSelectedItems());
  });

  it('<CSVLink> should have correct filename', async () => {
    jest.spyOn(reduxHooks, 'useSelector').mockReturnValue('7');

    render(<Flyout />);

    expect(screen.getByTestId('csvlink')).toHaveAttribute('download', '7_planets.csv');
  });
});
