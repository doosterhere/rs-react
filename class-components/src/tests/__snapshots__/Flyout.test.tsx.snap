// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flyout should render correctly 1`] = `
<div>
  <div
    class="flyout light"
    data-quantity="1"
  >
    <div>
      Selected items: 
      1
    </div>
    <button
      class="withIcon"
      role="button"
    >
      <i
        class="fa-solid fa-trash-can delete"
      />
      Unselect all
    </button>
    <a
      data-testid="csvlink"
      download="1_planet.csv"
      href="data:text/csv;charset=utf-8,ï»¿[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]
1"
      target="_self"
    >
      <button
        class="withIcon"
        role="button"
      >
        <i
          class="fa-solid fa-file-arrow-down"
        />
        Download selected
      </button>
    </a>
  </div>
</div>
`;
